multiple fixes

NkxxkN opened the pull request.

NkxxkN commented:
Fix for https://github.com/supertokens/supertokens-auth-react/issues/26

NkxxkN commented:
Fix for https://github.com/supertokens/supertokens-auth-react/issues/30

NkxxkN commented:
- Fix https://github.com/supertokens/supertokens-auth-react/issues/31
- Fix `onCallSignUpAPI` and `onCallSignInAPI` props not implementing FDI https://github.com/supertokens/frontend-driver-interface/blob/master/v1.3.0.md#output-9 correctly

NkxxkN commented on lib/.eslintrc.js:
```diff
@@ -112,5 +112,14 @@ module.exports = {
+          // disable the rule specifically for user input validators files.
+          "files": ["**/validators.ts"],
+          "rules": {
+            "@typescript-eslint/explicit-module-boundary-types": ["off"]
```
We want to allow `any` types for validators only.

NkxxkN commented:
Ready for review @rishabhpoddar

rishabhpoddar commented:
# IN PROGRESS
The review comments are above. A few more comments:
- I have not gone through `emailpassword/utils.ts` in too much detail. Just a skim. So please make sure the logic there is fine. One issue I think I found is that for password reset validators, whilst you use the password validator from sign up, you do not use the email validator from there (it's always equal to the default email validator, which is incorrect).

rishabhpoddar commented on lib/ts/recipe/session/session.ts:
```diff
@@ -125,6 +144,18 @@ export default class Session extends RecipeModule {
         return Session.getInstanceOrThrow().doesSessionExist();
     }
 
+    static addAxiosInterceptors(): boolean {
```
This takes an `axios` instance. The function signature in `supertokens-website` is:
```
addAxiosInterceptors: (axiosInstance: any) => void
```

rishabhpoddar commented on lib/ts/recipe/session/session.ts:
```diff
@@ -125,6 +144,18 @@ export default class Session extends RecipeModule {
+        return Session.getInstanceOrThrow().addAxiosInterceptors();
+    }
+
+    static setAuth0API(): boolean {
```
This also takes a param as in put.

rishabhpoddar commented on lib/ts/recipe/session/session.ts:
```diff
@@ -82,6 +89,18 @@ export default class Session extends RecipeModule {
+        return this.sessionSdk.setAuth0API();
+    };
+
+    getAuth0API = (): boolean => {
```
Incorrect return type

rishabhpoddar commented on lib/ts/recipe/session/session.ts:
```diff
@@ -82,6 +89,18 @@ export default class Session extends RecipeModule {
+        return this.sessionSdk.addAxiosInterceptors();
+    };
+
+    setAuth0API = (): boolean => {
```
Incorrect return type

rishabhpoddar commented on lib/ts/recipe/session/session.ts:
```diff
@@ -82,6 +89,18 @@ export default class Session extends RecipeModule {
         return this.sessionSdk.doesSessionExist();
     };
 
+    addAxiosInterceptors = (): boolean => {
```
Incorrect return type

rishabhpoddar commented on lib/ts/recipe/emailpassword/emailPassword.ts:
```diff
@@ -158,26 +158,23 @@ export default class EmailPassword extends RecipeModule {
 
     async signUpValidate(input: APIFormField[]): Promise<FormFieldError[]> {
-        return await validateFormOrThrow(input, this.config.signInAndUpFeature.signUpForm.formFields);
+        return await validateForm(input, this.config.signInAndUpFeature.signUpForm.formFields);
```
This can throw an error in case the input fields do not have the same length as the config form-field. In this case, what will happen?

Ideally, you should stick to only throwing an error in a function from a function or not throwing one at all. But OK.

NkxxkN replied:
The app would crash in that case. That should only happen while developers are developing a custom theme, trying to submit an incomplete form to the backend during development.
This is fairly important that the custom theme is implemented correctly and that this error is not being ignored so I was thinking that it's fine to throw unexpectedly here. Let me know if that's ok for you

rishabhpoddar commented on lib/ts/recipe/emailpassword/validators.ts:
```diff
@@ -0,0 +1,95 @@
+ * min 1 characters.
+ */
+
+export async function defaultLoginPasswordValidator(value: any): Promise<string | undefined> {
```
On the backend, I do not have this kind of a function. For sign in, password validation, I simply return `undefined`. The core separately checks if the password is not an empty string and if it is, it throws the wrong credentials error.

Would you like to have the same behaviour here? Or shall I change the backend?

NkxxkN replied:
I was under the impression that we decided to at least verify that there was one character but I didn't find any traces of this conversation. 

So I decided to check what big corps do:
- Facebook: no check for empty password
- Twitter: Disable the login button if empty password
- Instagram: Disable the login button if password is less than 6 characters

There is already a verification to make sure that all required fields are not empty:
![image](https://user-images.githubusercontent.com/5072452/98567473-40b31a00-22b0-11eb-877f-124b40da17d2.png)

Which makes the  `value.length === 0` check useless.

You don't need to implement the same thing in the backend, as you said, you can simply throw wrong credentials.

rishabhpoddar commented:
The review comments are above. A few more comments:
- I have not gone through `emailpassword/utils.ts` in too much detail. Just a skim. So please make sure the logic there is fine. One issue I think I found is that for password reset validators, whilst you use the password validator from sign up, you do not use the email validator from there (it's always equal to the default email validator, which is incorrect).
- Please confirm that you have seen, applied and understood the unchecked item at the bottom [this comment](https://github.com/supertokens/supertokens-auth-react/issues/27#issuecomment-722900699) 
- The three new functions added in the session recipe are not being exposed via the `session/index.ts` file
- There are still occurrences of implicit checks with props in the Components. You do not need to change the code for this now since it probably won't be solved quickly.

NkxxkN commented on lib/ts/recipe/emailpassword/components/signInAndUp/api.ts:
```diff
@@ -0,0 +1,113 @@
+        return {
+            status: API_RESPONSE_STATUS.GENERAL_ERROR,
+            message: SOMETHING_WENT_WRONG_ERROR
+        };
```
Right, I wasn't sure if the API was returning `GENERAL_ERROR` in some cases or if it was only for status code > 300.
I've refactored so that API response type does not include GENERAL_ERROR and theme response types includes it.

rishabhpoddar replied:
According to the FDI spec, there should never be a status of `GENERAL_ERROR` at all returned by the API.

NkxxkN commented:
> whilst you use the password validator from sign up, you do not use the email validator from there (it's always equal to the default email validator, which is incorrect).

Fixed

> Please confirm that you have seen, applied and understood the unchecked item at the bottom this comment

Done

> The three new functions added in the session recipe are not being exposed via the session/index.ts file

Fixed

> There are still occurrences of implicit checks with props in the Components. You do not need to change the code for this now since it probably won't be solved quickly.

Fixed

rishabhpoddar commented on lib/ts/recipe/emailpassword/components/signInAndUp/SignInAndUp.tsx:
```diff
@@ -180,67 +123,35 @@ class SignInAndUp extends Component<SignInAndUpProps, { isLoading: boolean; user
-                }
+    setStateOnSuccessfulAPICall(normalisedAPIResponse: SignInThemeResponse | SignUpThemeResponse): void {
+        let user: User;
+        if (normalisedAPIResponse.status === API_RESPONSE_STATUS.OK && normalisedAPIResponse.user) {
```
@NkxxkN this comment has not be resolved yet.
